# Titanic Data Analysis 

Frederick Ferguson
February 22, 2018

Please see answers at the end of the notebook

# Data Dictionary
Variable	Definition	    Key
survival	Survival	    0 = No, 1 = Yes
pclass	    Ticket class	1 = 1st, 2 = 2nd, 3 = 3rd
sex	        Sex	
Age	        Age in years	
sibsp	    # of siblings / spouses aboard the Titanic	
parch	    # of parents / children aboard the Titanic	
ticket	    Ticket number	
fare	    Passenger fare	
cabin	    Cabin number	
embarked	Port of Embarkation	C = Cherbourg, Q = Queenstown, S = Southampton

In [221]:
%matplotlib inline
import numpy as np
import pandas as pd
from pandas import DataFrame, Series
import matplotlib.pyplot as plt

df=pd.read_csv('Assignment_2_Titanic_data_train.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
5              6         0       3   
6              7         0       1   
7              8         0       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
12            13         0       3   
13            14         0       3   
14            15         0       3   
15            16         1       2   
16            17         0       3   
17            18         1       2   
18            19         0       3   
19            20         1       3   
20            21         0       2   
21            22         1       2   
22            23         1       3   
23            24         1       1   
24            25         0       3   
25            26         1       3   
26            27         0       3   
27            28         0       1   
28            29         1       3   
29            30         0       3   
..           ...       ...     ...   
861          862         0       2   
862          863         1       1   
863          864         0       3   
864          865         0       2   
865          866         1       2   
866          867         1       2   
867          868         0       1   
868          869         0       3   
869          870         1       3   
870          871         0       3   
871          872         1       1   
872          873         0       1   
873          874         0       3   
874          875         1       2   
875          876         1       3   
876          877         0       3   
877          878         0       3   
878          879         0       3   
879          880         1       1   
880          881         1       2   
881          882         0       3   
882          883         0       3   
883          884         0       2   
884          885         0       3   
885          886         0       3   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
12                      Saundercock, Mr. William Henry    male  20.0      0   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                                Rice, Master. Eugene    male   2.0      4   
17                        Williams, Mr. Charles Eugene    male   NaN      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN

In [222]:
# Inspect datatypes
df.dtypes 

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [223]:
#Understand why age is a float
df.Age.unique()

array([ 22.  ,  38.  ,  26.  ,  35.  ,    nan,  54.  ,   2.  ,  27.  ,
        14.  ,   4.  ,  58.  ,  20.  ,  39.  ,  55.  ,  31.  ,  34.  ,
        15.  ,  28.  ,   8.  ,  19.  ,  40.  ,  66.  ,  42.  ,  21.  ,
        18.  ,   3.  ,   7.  ,  49.  ,  29.  ,  65.  ,  28.5 ,   5.  ,
        11.  ,  45.  ,  17.  ,  32.  ,  16.  ,  25.  ,   0.83,  30.  ,
        33.  ,  23.  ,  24.  ,  46.  ,  59.  ,  71.  ,  37.  ,  47.  ,
        14.5 ,  70.5 ,  32.5 ,  12.  ,   9.  ,  36.5 ,  51.  ,  55.5 ,
        40.5 ,  44.  ,   1.  ,  61.  ,  56.  ,  50.  ,  36.  ,  45.5 ,
        20.5 ,  62.  ,  41.  ,  52.  ,  63.  ,  23.5 ,   0.92,  43.  ,
        60.  ,  10.  ,  64.  ,  13.  ,  48.  ,   0.75,  53.  ,  57.  ,
        80.  ,  70.  ,  24.5 ,   6.  ,   0.67,  30.5 ,   0.42,  34.5 ,  74.  ])

In [224]:
#  Determine sparsity
obs=len(df)
print("Number of observations: %d" % obs)
cols=df.columns.values.tolist()
for col in cols:
    nn_cnt=df[col].count()
    nn_pct=nn_cnt/obs*100
    unq_val=len(df[col].unique())
    print("    %s:" % col)
    print("       Number of not_null: %d or %.1f%%" % (nn_cnt,nn_pct))
    print("       Number of unique values: %d" % unq_val)

Number of observations: 891
    PassengerId:
       Number of not_null: 891 or 100.0%
       Number of unique values: 891
    Survived:
       Number of not_null: 891 or 100.0%
       Number of unique values: 2
    Pclass:
       Number of not_null: 891 or 100.0%
       Number of unique values: 3
    Name:
       Number of not_null: 891 or 100.0%
       Number of unique values: 891
    Sex:
       Number of not_null: 891 or 100.0%
       Number of unique values: 2
    Age:
       Number of not_null: 714 or 80.1%
       Number of unique values: 89
    SibSp:
       Number of not_null: 891 or 100.0%
       Number of unique values: 7
    Parch:
       Number of not_null: 891 or 100.0%
       Number of unique values: 7
    Ticket:
       Number of not_null: 891 or 100.0%
       Number of unique values: 681
    Fare:
       Number of not_null: 891 or 100.0%
       Number of unique values: 248
    Cabin:
       Number of not_null: 204 or 22.9%
       Number of unique values: 148
    Embarked

In [225]:
# Drop columns
del df['Name']   #all unique--not explanatory
del df['Ticket'] #mostly unique--not explantory
del df['Cabin']  #sparse
df

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch      Fare  \
0              1         0       3    male  22.0      1      0    7.2500   
1              2         1       1  female  38.0      1      0   71.2833   
2              3         1       3  female  26.0      0      0    7.9250   
3              4         1       1  female  35.0      1      0   53.1000   
4              5         0       3    male  35.0      0      0    8.0500   
5              6         0       3    male   NaN      0      0    8.4583   
6              7         0       1    male  54.0      0      0   51.8625   
7              8         0       3    male   2.0      3      1   21.0750   
8              9         1       3  female  27.0      0      2   11.1333   
9             10         1       2  female  14.0      1      0   30.0708   
10            11         1       3  female   4.0      1      1   16.7000   
11            12         1       1  female  58.0      0      0   26.5500   
12            13         0       3    male  20.0      0      0    8.0500   
13            14         0       3    male  39.0      1      5   31.2750   
14            15         0       3  female  14.0      0      0    7.8542   
15            16         1       2  female  55.0      0      0   16.0000   
16            17         0       3    male   2.0      4      1   29.1250   
17            18         1       2    male   NaN      0      0   13.0000   
18            19         0       3  female  31.0      1      0   18.0000   
19            20         1       3  female   NaN      0      0    7.2250   
20            21         0       2    male  35.0      0      0   26.0000   
21            22         1       2    male  34.0      0      0   13.0000   
22            23         1       3  female  15.0      0      0    8.0292   
23            24         1       1    male  28.0      0      0   35.5000   
24            25         0       3  female   8.0      3      1   21.0750   
25            26         1       3  female  38.0      1      5   31.3875   
26            27         0       3    male   NaN      0      0    7.2250   
27            28         0       1    male  19.0      3      2  263.0000   
28            29         1       3  female   NaN      0      0    7.8792   
29            30         0       3    male   NaN      0      0    7.8958   
..           ...       ...     ...     ...   ...    ...    ...       ...   
861          862         0       2    male  21.0      1      0   11.5000   
862          863         1       1  female  48.0      0      0   25.9292   
863          864         0       3  female   NaN      8      2   69.5500   
864          865         0       2    male  24.0      0      0   13.0000   
865          866         1       2  female  42.0      0      0   13.0000   
866          867         1       2  female  27.0      1      0   13.8583   
867          868         0       1    male  31.0      0      0   50.4958   
868          869         0       3    male   NaN      0      0    9.5000   
869          870         1       3    male   4.0      1      1   11.1333   
870          871         0       3    male  26.0      0      0    7.8958   
871          872         1       1  female  47.0      1      1   52.5542   
872          873         0       1    male  33.0      0      0    5.0000   
873          874         0       3    male  47.0      0      0    9.0000   
874          875         1       2  female  28.0      1      0   24.0000   
875          876         1       3  female  15.0      0      0    7.2250   
876          877         0       3    male  20.0      0      0    9.8458   
877          878         0       3    male  19.0      0      0    7.8958   
878          879         0       3    male   NaN      0      0    7.8958   
879          880         1       1  female  56.0      0      1   83.1583   
880          881         1       2  female  25.0      0      1   26.0000   
881          882         0       3    male  33.0     

In [226]:
#Look at Sex value counts
sex_counts=df.Sex.value_counts()
print(sex_counts)
males=sex_counts.iloc[0]
females=sex_counts.iloc[0]

male      577
female    314
Name: Sex, dtype: int64


In [227]:
#Look at Sex percentages
df.Sex.value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

There were twice as many men as women.

In [228]:
#Determine average age by gender
aabg=df.groupby('Sex', as_index=False).Age.mean()
#print(aabg)
faa=aabg.iloc[0]['Age']
maa=aabg.iloc[1]['Age']
print("Female avg. age: %1f" % faa)
print("Male avg. age: %1f" % maa)

Female avg. age: 27.915709
Male avg. age: 30.726645


The women were slightly younger than the men.

In [229]:
# Replace NaN Age with mean by Sex
df["Age"].fillna(df.groupby("Sex")["Age"].transform("mean"), inplace=True)
print(df)

     PassengerId  Survived  Pclass     Sex        Age  SibSp  Parch      Fare  \
0              1         0       3    male  22.000000      1      0    7.2500   
1              2         1       1  female  38.000000      1      0   71.2833   
2              3         1       3  female  26.000000      0      0    7.9250   
3              4         1       1  female  35.000000      1      0   53.1000   
4              5         0       3    male  35.000000      0      0    8.0500   
5              6         0       3    male  30.726645      0      0    8.4583   
6              7         0       1    male  54.000000      0      0   51.8625   
7              8         0       3    male   2.000000      3      1   21.0750   
8              9         1       3  female  27.000000      0      2   11.1333   
9             10         1       2  female  14.000000      1      0   30.0708   
10            11         1       3  female   4.000000      1      1   16.7000   
11            12         1  

In [230]:
#make a survivors dataframe
dfs=df.loc[df['Survived'] == 1]
survivors=len(dfs)
print("Number of Survivors: %d" % survivors)

Number of Survivors: 342


In [231]:
#analysis survivors by gender
survived_by_gender=dfs.groupby('Sex', as_index=False).PassengerId.count()
print(survived_by_gender)
male_surv=survived_by_gender.iloc[1]["PassengerId"]
female_surv=survived_by_gender.iloc[0]["PassengerId"]
msp=male_surv/males*100
fsp=female_surv/females*100
print("males survived: %d%%" % msp)
print("females survived: %d%%" % fsp)

      Sex  PassengerId
0  female          233
1    male          109
males survived: 18%
females survived: 40%


Women survived at more than double the rate of men.  Clearly gender is an explantory variable.

In [232]:
dfs['Age'].groupby(dfs['Sex']).describe()

count       mean        std   min   25%        50%   75%   max
Sex                                                                   
female  233.0  28.703715  13.033320  0.75  22.0  27.915709  35.0  63.0
male    109.0  27.782535  15.282563  0.42  20.0  30.726645  35.0  80.0

The avg female age was 27.9 but for survivors it was 28.7
The avg male age was 30.726645 but for survivors it was 27.8 or lower than for females.  
Being younger helped men but not women.

In [233]:
dfs['Pclass'].describe()

count    342.000000
mean       1.950292
std        0.863321
min        1.000000
25%        1.000000
50%        2.000000
75%        3.000000
max        3.000000
Name: Pclass, dtype: float64

In [234]:
#Determine survivors average Pclass by gender
apcbg=dfs.groupby('Sex', as_index=False).Pclass.mean()
print(apcbg)
fapc=apcbg.iloc[0]['Pclass']
mapc=apcbg.iloc[1]['Pclass']
print("Female survivor avg. Pclass: %1f" % fapc)
print("Male survivor avg. Pclass: %1f" % mapc)

      Sex    Pclass
0  female  1.918455
1    male  2.018349
Female survivor avg. Pclass: 1.918455
Male survivor avg. Pclass: 2.018349


It looks like Pclass is a weak explanatory variable but 1st or 2nd class seems better for survival.

In [235]:
#Determine survivors SibSip
dfs.SibSp.value_counts(normalize=True)

0    0.614035
1    0.327485
2    0.038012
3    0.011696
4    0.008772
Name: SibSp, dtype: float64

No Siblings looks better

In [236]:
#Determine survivors Parch
dfs.Parch.value_counts(normalize=True)

0    0.681287
1    0.190058
2    0.116959
3    0.008772
5    0.002924
Name: Parch, dtype: float64

No Parents looks better

In [237]:
#Determine survivors avg. fare
print("The average fare was %f" % df.Fare.mean())
print("The average survivors fare was %f" % dfs.Fare.mean())

The average fare was 32.204208
The average survivors fare was 48.395408


Definitely a higher fare helped.

In [238]:
#Count Overall Embarked
df.Embarked.value_counts(normalize=True)

S    0.724409
C    0.188976
Q    0.086614
Name: Embarked, dtype: float64

In [239]:
#Count Survivors Embarked
dfs.Embarked.value_counts(normalize=True)

S    0.638235
C    0.273529
Q    0.088235
Name: Embarked, dtype: float64

Embarking from Cherbourg seems to have helped.

# Let's try building a Gradient Boosting and Random Forest in H2O

In [268]:
import h2o
h2o.init(max_mem_size="2G")
h2o.remove_all()

from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.estimators.random_forest import H2ORandomForestEstimator
from h2o.model.metrics_base import H2OMultinomialModelMetrics
from h2o.estimators.glm import G



Checking whether there is an H2O instance running at http://localhost:54321. connected.


--------------------------  ----------------------------------------
H2O cluster uptime:         1 day 2 hours 40 mins
H2O cluster timezone:       America/Toronto
H2O data parsing timezone:  UTC
H2O cluster version:        3.18.0.1
H2O cluster version age:    6 days
H2O cluster name:           aliennationrefugeee
H2O cluster total nodes:    1
H2O cluster free memory:    1.613 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         locked, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.4 final
--------------------------  ----------------------------------------

We import the reduced Titanic dataset (891k rows, 9 columns, 4 numerical, 5 categorical) and then split the data 3 ways:  
    70% for training  
    20% for validation (hyper parameter tuning)  
    10% for final testing  

We will train a data set on one set and use the others to test the validity of the model by ensuring that it can predict accurately on data the model has not been shown.  The second set will be used for validation most of the time.  The third set will be withheld until the end, to ensure that our validation accuracy is consistent with data we have never seen during the iterative process. 

In [241]:
#Import the Titanic dataframe into an H2O dataframe
hf = h2o.H2OFrame(df)

#Reorder columns with response variable at the end dropping PassengerId
hf=hf[['Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare','Survived']]
hf

Parse progress: |█████████████████████████████████████████████████████████| 100%


Pclass Sex Age SibSp Parch Fare Survived 
 
 
 3 male 22 1 0 7.25 0 
 1 female 38 1 0 71.2833 1 
 3 female 26 0 0 7.925 1 
 1 female 35 1 0 53.1 1 
 3 male 35 0 0 8.05 0 
 3 male 30.7266 0 0 8.4583 0 
 1 male 54 0 0 51.8625 0 
 3 male 2 3 1 21.075 0 
 3 female 27 0 2 11.1333 1 
 2 female 14 1 0 30.0708 1

In [242]:
#split out training, validation and testing frames (70/20/10)
trainf, validf, testf=hf.split_frame([.7,.2],seed=314159)

In [244]:
#Prepare explanatory columns
hf_X=hf.col_names[:-1]
hf_Y=hf.col_names[-1]

Build Random Forest model with the following parameters:
 
 **model_id:** Not required, but allows us to easily find our model in the [Flow](http://localhost:54321/) interface  
 **ntrees:** Maximum number of trees used by the random forest. Default value is 50. We can afford to increase this, as our early-stopping criterion will decide when the random forest is sufficiently accurate.  
 **stopping_rounds:** Stopping criterion described above. Stops fitting new trees when 2-tree rolling average is within 0.001 (default) of the two prior rolling averages. Can be thought of as a convergence setting.  
 **score_each_teration:** predict against training and validation for each tree. Default will skip several.  
 **seed:** set the randomization seed so we can reproduce results

In [254]:
#Random Forest Model Parameters
rf_Titanic_v1 = H2ORandomForestEstimator(
    model_id="rf_Titanic_v1",
    ntrees=200,
    stopping_rounds=2,
    score_each_iteration=True,
    seed=1000000)

In [255]:
#Random Forest Model Construction
rf_Titanic_v1.train(hf_X, hf_Y, training_frame=trainf, validation_frame=validf)

drf Model Build progress: |███████████████████████████████████████████████| 100%


In [256]:
#Take a look at the RF model overview
rf_Titanic_v1

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  rf_Titanic_v1


ModelMetricsRegression: drf
** Reported on train data. **

MSE: 0.15979280795813916
RMSE: 0.39974092604853356
MAE: 0.24597266212759553
RMSLE: 0.2801454671450774
Mean Residual Deviance: 0.15979280795813916

ModelMetricsRegression: drf
** Reported on validation data. **

MSE: 0.15841079259746266
RMSE: 0.3980085333224184
MAE: 0.2665473800675496
RMSLE: 0.2871604515614969
Mean Residual Deviance: 0.15841079259746266
Scoring History: 


timestamp            duration    number_of_trees    training_rmse    training_mae    training_deviance    validation_rmse    validation_mae    validation_deviance
--  -------------------  ----------  -----------------  ---------------  --------------  -------------------  -----------------  ----------------  ---------------------
    2018-02-19 21:43:00  0.036 sec   0                  nan              nan             nan                  nan                nan               nan
    2018-02-19 21:43:00  0.096 sec   1                  0.49906          0.259285        0.24906              0.480611           0.24828           0.230987
    2018-02-19 21:43:00  0.143 sec   2                  0.48205          0.259872        0.232372             0.451903           0.26733           0.204216
    2018-02-19 21:43:00  0.150 sec   3                  0.476122         0.262888        0.226692             0.438147           0.263512          0.191973
    2018-02-19 21:43:00  0.159 sec   4                  0.457733         0.253604        0.20952              0.434605           0.267287          0.188882
    2018-02-19 21:43:00  0.165 sec   5                  0.443385         0.247136        0.19659              0.426245           0.26544           0.181685
    2018-02-19 21:43:00  0.171 sec   6                  0.43169          0.243247        0.186357             0.417624           0.264231          0.17441
    2018-02-19 21:43:00  0.179 sec   7                  0.424978         0.243325        0.180606             0.398156           0.253239          0.158528
    2018-02-19 21:43:00  0.185 sec   8                  0.42023          0.245156        0.176593             0.398505           0.256025          0.158806
    2018-02-19 21:43:00  0.198 sec   9                  0.416395         0.244134        0.173385             0.393663           0.256289          0.15497
    2018-02-19 21:43:00  0.209 sec   10                 0.407976         0.240285        0.166444             0.397408           0.259947          0.157933
    2018-02-19 21:43:00  0.216 sec   11                 0.403628         0.245163        0.162915             0.396133           0.262481          0.156922
    2018-02-19 21:43:00  0.226 sec   12                 0.400033         0.243591        0.160027             0.393961           0.260727          0.155205
    2018-02-19 21:43:00  0.233 sec   13                 0.398543         0.244144        0.158837             0.398323           0.266449          0.158661
    2018-02-19 21:43:00  0.257 sec   14                 0.399741         0.245973        0.159793             0.398009           0.266547          0.158411

Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
Sex         463.22                 1                    0.339036
Fare        339.741                0.733434             0.248661
Age         311.986                0.673515             0.228346
Pclass      136.659                0.295021             0.100023
SibSp       59.5138                0.128479             0.0435589
Parch       55.1644                0.119089             0.0403755

The most important explanatory variables for survival are: 1) Sex, 2) Age, 3) Fare and 4) Pclass

In [257]:
#Take a look at the RF validation statistics
rf_Titanic_v1.score_history()

timestamp    duration  number_of_trees  training_rmse  \
0     2018-02-19 21:43:00   0.036 sec              0.0            NaN   
1     2018-02-19 21:43:00   0.096 sec              1.0       0.499060   
2     2018-02-19 21:43:00   0.143 sec              2.0       0.482050   
3     2018-02-19 21:43:00   0.150 sec              3.0       0.476122   
4     2018-02-19 21:43:00   0.159 sec              4.0       0.457733   
5     2018-02-19 21:43:00   0.165 sec              5.0       0.443385   
6     2018-02-19 21:43:00   0.171 sec              6.0       0.431690   
7     2018-02-19 21:43:00   0.179 sec              7.0       0.424978   
8     2018-02-19 21:43:00   0.185 sec              8.0       0.420230   
9     2018-02-19 21:43:00   0.198 sec              9.0       0.416395   
10    2018-02-19 21:43:00   0.209 sec             10.0       0.407976   
11    2018-02-19 21:43:00   0.216 sec             11.0       0.403628   
12    2018-02-19 21:43:00   0.226 sec             12.0       0.400033   
13    2018-02-19 21:43:00   0.233 sec             13.0       0.398543   
14    2018-02-19 21:43:00   0.257 sec             14.0       0.399741   

    training_mae  training_deviance  validation_rmse  validation_mae  \
0            NaN                NaN              NaN             NaN   
1       0.259285           0.249060         0.480611        0.248280   
2       0.259872           0.232372         0.451903        0.267330   
3       0.262888           0.226692         0.438147        0.263512   
4       0.253604           0.209520         0.434605        0.267287   
5       0.247136           0.196590         0.426245        0.265440   
6       0.243247           0.186357         0.417624        0.264231   
7       0.243325           0.180606         0.398156        0.253239   
8       0.245156           0.176593         0.398505        0.256025   
9       0.244134           0.173385         0.393663        0.256289   
10      0.240285           0.166444         0.397408        0.259947   
11      0.245163           0.162915         0.396133        0.262481   
12      0.243591           0.160027         0.393961        0.260727   
13      0.244144           0.158837         0.398323        0.266449   
14      0.245973           0.159793         0.398009        0.266547   

    validation_deviance  
0                   NaN  
1              0.230987  
2              0.204216  
3              0.191973  
4              0.188882  
5              0.181685  
6              0.174410  
7              0.158528  
8              0.158806  
9              0.154970  
10             0.157933  
11             0.156922  
12             0.155205  
13             0.158661  
14             0.158411

In [284]:
rf_Titanic_v1

Model Details
H2ORandomForestEstimator :  Distributed Random Forest
Model Key:  rf_Titanic_v1


ModelMetricsRegression: drf
** Reported on train data. **

MSE: 0.15979280795813916
RMSE: 0.39974092604853356
MAE: 0.24597266212759553
RMSLE: 0.2801454671450774
Mean Residual Deviance: 0.15979280795813916

ModelMetricsRegression: drf
** Reported on validation data. **

MSE: 0.15841079259746266
RMSE: 0.3980085333224184
MAE: 0.2665473800675496
RMSLE: 0.2871604515614969
Mean Residual Deviance: 0.15841079259746266
Scoring History: 


timestamp            duration    number_of_trees    training_rmse    training_mae    training_deviance    validation_rmse    validation_mae    validation_deviance
--  -------------------  ----------  -----------------  ---------------  --------------  -------------------  -----------------  ----------------  ---------------------
    2018-02-19 21:43:00  0.036 sec   0                  nan              nan             nan                  nan                nan               nan
    2018-02-19 21:43:00  0.096 sec   1                  0.49906          0.259285        0.24906              0.480611           0.24828           0.230987
    2018-02-19 21:43:00  0.143 sec   2                  0.48205          0.259872        0.232372             0.451903           0.26733           0.204216
    2018-02-19 21:43:00  0.150 sec   3                  0.476122         0.262888        0.226692             0.438147           0.263512          0.191973
    2018-02-19 21:43:00  0.159 sec   4                  0.457733         0.253604        0.20952              0.434605           0.267287          0.188882
    2018-02-19 21:43:00  0.165 sec   5                  0.443385         0.247136        0.19659              0.426245           0.26544           0.181685
    2018-02-19 21:43:00  0.171 sec   6                  0.43169          0.243247        0.186357             0.417624           0.264231          0.17441
    2018-02-19 21:43:00  0.179 sec   7                  0.424978         0.243325        0.180606             0.398156           0.253239          0.158528
    2018-02-19 21:43:00  0.185 sec   8                  0.42023          0.245156        0.176593             0.398505           0.256025          0.158806
    2018-02-19 21:43:00  0.198 sec   9                  0.416395         0.244134        0.173385             0.393663           0.256289          0.15497
    2018-02-19 21:43:00  0.209 sec   10                 0.407976         0.240285        0.166444             0.397408           0.259947          0.157933
    2018-02-19 21:43:00  0.216 sec   11                 0.403628         0.245163        0.162915             0.396133           0.262481          0.156922
    2018-02-19 21:43:00  0.226 sec   12                 0.400033         0.243591        0.160027             0.393961           0.260727          0.155205
    2018-02-19 21:43:00  0.233 sec   13                 0.398543         0.244144        0.158837             0.398323           0.266449          0.158661
    2018-02-19 21:43:00  0.257 sec   14                 0.399741         0.245973        0.159793             0.398009           0.266547          0.158411

Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
Sex         463.22                 1                    0.339036
Fare        339.741                0.733434             0.248661
Age         311.986                0.673515             0.228346
Pclass      136.659                0.295021             0.100023
SibSp       59.5138                0.128479             0.0435589
Parch       55.1644                0.119089             0.0403755

Build Gradient Boosting model with the following parameters:
 
 **model_id:** Not required, but allows us to easily find our model in the [Flow](http://localhost:54321/) interface  
 **seed:** set the randomization seed so we can reproduce results

In [260]:
gbm_Titanic_v1 = H2OGradientBoostingEstimator(
    model_id="gbm_Titanic_v1",
    seed=2000000
)
gbm_Titanic_v1.train(hf_X, hf_Y, training_frame=trainf, validation_frame=validf)

gbm Model Build progress: |███████████████████████████████████████████████| 100%


In [261]:
#Take a look at the GBM model overview
gbm_Titanic_v1

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  gbm_Titanic_v1


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 0.08678955462848892
RMSE: 0.294600669769247
MAE: 0.2079455915103304
RMSLE: 0.20694048882364288
Mean Residual Deviance: 0.08678955462848892

ModelMetricsRegression: gbm
** Reported on validation data. **

MSE: 0.14959868503549564
RMSE: 0.3867798922326439
MAE: 0.27914018441979865
RMSLE: 0.27721168926215844
Mean Residual Deviance: 0.14959868503549564
Scoring History: 


timestamp            duration    number_of_trees    training_rmse        training_mae         training_deviance    validation_rmse      validation_mae       validation_deviance
---  -------------------  ----------  -----------------  -------------------  -------------------  -------------------  -------------------  -------------------  ---------------------
     2018-02-19 21:44:14  0.028 sec   0.0                0.48963364981452917  0.47948222206139207  0.239741111030697    0.47755427159940084  0.4677991933535306   0.2280580823228343
     2018-02-19 21:44:14  0.082 sec   1.0                0.4652766909617107   0.4549993673800655   0.21648239915227926  0.45714645337026927  0.44719710529801643  0.20898287982901576
     2018-02-19 21:44:14  0.094 sec   2.0                0.4445664664865269   0.4329834410806944   0.19763934312431625  0.44147498330057716  0.43006415105805273  0.19490016088024487
     2018-02-19 21:44:14  0.100 sec   3.0                0.42699407424423447  0.41315895884366666  0.18232393943969083  0.42745577545372876  0.41363218272189217  0.1827184399687486
     2018-02-19 21:44:14  0.106 sec   4.0                0.4117913900407737   0.3949496288966519   0.1695721489117126   0.4169804225226581   0.399614945589574    0.17387267276717447
---  ---                  ---         ---                ---                  ---                  ---                  ---                  ---                  ---
     2018-02-19 21:44:14  0.263 sec   46.0               0.29680929039056325  0.21024920192616017  0.08809575486214971  0.38474690000840095  0.27860889972472325  0.14803017706607446
     2018-02-19 21:44:14  0.267 sec   47.0               0.29651615826005556  0.20993846065778032  0.08792183210930232  0.3852553152419986   0.2787791703891852   0.14842165792221168
     2018-02-19 21:44:14  0.270 sec   48.0               0.2949826446396572   0.2083596730192261   0.08701476063860628  0.3862569542491001   0.27909078504675755  0.1491944347057914
     2018-02-19 21:44:14  0.274 sec   49.0               0.2948383247289026   0.2081874007568021   0.08692963772894581  0.3862932902543789   0.27898908207325795  0.14922250609555385
     2018-02-19 21:44:14  0.278 sec   50.0               0.294600669769247    0.2079455915103304   0.08678955462848892  0.3867798922326439   0.27914018441979865  0.14959868503549564


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
Sex         234.053                1                    0.467437
Age         92.302                 0.394364             0.18434
Fare        83.0249                0.354727             0.165813
Pclass      63.8294                0.272713             0.127476
SibSp       20.9053                0.0893188            0.0417509
Parch       6.60053                0.028201             0.0131822

In [262]:
#Take a look at GBM validation statistics
gbm_Titanic_v1.score_history()

timestamp    duration  number_of_trees  training_rmse  \
0     2018-02-19 21:44:14   0.028 sec              0.0       0.489634   
1     2018-02-19 21:44:14   0.082 sec              1.0       0.465277   
2     2018-02-19 21:44:14   0.094 sec              2.0       0.444566   
3     2018-02-19 21:44:14   0.100 sec              3.0       0.426994   
4     2018-02-19 21:44:14   0.106 sec              4.0       0.411791   
5     2018-02-19 21:44:14   0.109 sec              5.0       0.398238   
6     2018-02-19 21:44:14   0.113 sec              6.0       0.386791   
7     2018-02-19 21:44:14   0.124 sec              7.0       0.377045   
8     2018-02-19 21:44:14   0.131 sec              8.0       0.368195   
9     2018-02-19 21:44:14   0.134 sec              9.0       0.360057   
10    2018-02-19 21:44:14   0.138 sec             10.0       0.353287   
11    2018-02-19 21:44:14   0.141 sec             11.0       0.348182   
12    2018-02-19 21:44:14   0.145 sec             12.0       0.343453   
13    2018-02-19 21:44:14   0.148 sec             13.0       0.339437   
14    2018-02-19 21:44:14   0.152 sec             14.0       0.335473   
15    2018-02-19 21:44:14   0.154 sec             15.0       0.331770   
16    2018-02-19 21:44:14   0.157 sec             16.0       0.329376   
17    2018-02-19 21:44:14   0.160 sec             17.0       0.326867   
18    2018-02-19 21:44:14   0.162 sec             18.0       0.324325   
19    2018-02-19 21:44:14   0.165 sec             19.0       0.321218   
20    2018-02-19 21:44:14   0.168 sec             20.0       0.319998   
21    2018-02-19 21:44:14   0.175 sec             21.0       0.318104   
22    2018-02-19 21:44:14   0.178 sec             22.0       0.315544   
23    2018-02-19 21:44:14   0.181 sec             23.0       0.313521   
24    2018-02-19 21:44:14   0.185 sec             24.0       0.311793   
25    2018-02-19 21:44:14   0.189 sec             25.0       0.309897   
26    2018-02-19 21:44:14   0.192 sec             26.0       0.308220   
27    2018-02-19 21:44:14   0.195 sec             27.0       0.306863   
28    2018-02-19 21:44:14   0.199 sec             28.0       0.306324   
29    2018-02-19 21:44:14   0.202 sec             29.0       0.305574   
30    2018-02-19 21:44:14   0.205 sec             30.0       0.305140   
31    2018-02-19 21:44:14   0.208 sec             31.0       0.304573   
32    2018-02-19 21:44:14   0.211 sec             32.0       0.302909   
33    2018-02-19 21:44:14   0.213 sec             33.0       0.302568   
34    2018-02-19 21:44:14   0.217 sec             34.0       0.301470   
35    2018-02-19 21:44:14   0.219 sec             35.0       0.301218   
36    2018-02-19 21:44:14   0.222 sec             36.0       0.300960   
37    2018-02-19 21:44:14   0.226 sec             37.0       0.300586   
38    2018-02-19 21:44:14   0.233 sec             38.0       0.300262   
39    2018-02-19 21:44:14   0.237 sec             39.0       0.300065   
40    2018-02-19 21:44:14   0.241 sec             40.0       0.299124   
41    2018-02-19 21:44:14   0.244 sec             41.0       0.297912   
42    2018-02-19 21:44:14   0.248 sec             42.0       0.297739   
43    2018-02-19 21:44:14   0.251 sec             43.0       0.297528   
44    2018-02-19 21:44:14   0.253 sec             44.0       0.297386   
45    2018-02-19 21:44:14   0.257 sec             45.0       0.297220   
46    2018-02-19 21:44:14   0.263 sec             46.0       0.296809   
47    2018-02-19 21:44:14   0.267 sec             47.0       0.296516   
48    2018-02-19 21:44:14   0.270 sec             48.0       0.294983   
49    2018-02-19 21:44:14   0.274 sec             49.0       0.294838   
50    2018-02-19 21:44:14   0.278 sec             50.0       0.294601   

    training_mae  training_deviance  validation_rmse  validation_mae  \
0       0.479482           0.239741         0.477554        0.467799   
1       0.454999           0.216482         0.457146        0.447197   
2  

In [283]:
gbm_Titanic_v1


Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  gbm_Titanic_v1


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 0.08678955462848892
RMSE: 0.294600669769247
MAE: 0.2079455915103304
RMSLE: 0.20694048882364288
Mean Residual Deviance: 0.08678955462848892

ModelMetricsRegression: gbm
** Reported on validation data. **

MSE: 0.14959868503549564
RMSE: 0.3867798922326439
MAE: 0.27914018441979865
RMSLE: 0.27721168926215844
Mean Residual Deviance: 0.14959868503549564
Scoring History: 


timestamp            duration    number_of_trees    training_rmse        training_mae         training_deviance    validation_rmse      validation_mae       validation_deviance
---  -------------------  ----------  -----------------  -------------------  -------------------  -------------------  -------------------  -------------------  ---------------------
     2018-02-19 21:44:14  0.028 sec   0.0                0.48963364981452917  0.47948222206139207  0.239741111030697    0.47755427159940084  0.4677991933535306   0.2280580823228343
     2018-02-19 21:44:14  0.082 sec   1.0                0.4652766909617107   0.4549993673800655   0.21648239915227926  0.45714645337026927  0.44719710529801643  0.20898287982901576
     2018-02-19 21:44:14  0.094 sec   2.0                0.4445664664865269   0.4329834410806944   0.19763934312431625  0.44147498330057716  0.43006415105805273  0.19490016088024487
     2018-02-19 21:44:14  0.100 sec   3.0                0.42699407424423447  0.41315895884366666  0.18232393943969083  0.42745577545372876  0.41363218272189217  0.1827184399687486
     2018-02-19 21:44:14  0.106 sec   4.0                0.4117913900407737   0.3949496288966519   0.1695721489117126   0.4169804225226581   0.399614945589574    0.17387267276717447
---  ---                  ---         ---                ---                  ---                  ---                  ---                  ---                  ---
     2018-02-19 21:44:14  0.263 sec   46.0               0.29680929039056325  0.21024920192616017  0.08809575486214971  0.38474690000840095  0.27860889972472325  0.14803017706607446
     2018-02-19 21:44:14  0.267 sec   47.0               0.29651615826005556  0.20993846065778032  0.08792183210930232  0.3852553152419986   0.2787791703891852   0.14842165792221168
     2018-02-19 21:44:14  0.270 sec   48.0               0.2949826446396572   0.2083596730192261   0.08701476063860628  0.3862569542491001   0.27909078504675755  0.1491944347057914
     2018-02-19 21:44:14  0.274 sec   49.0               0.2948383247289026   0.2081874007568021   0.08692963772894581  0.3862932902543789   0.27898908207325795  0.14922250609555385
     2018-02-19 21:44:14  0.278 sec   50.0               0.294600669769247    0.2079455915103304   0.08678955462848892  0.3867798922326439   0.27914018441979865  0.14959868503549564


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
Sex         234.053                1                    0.467437
Age         92.302                 0.394364             0.18434
Fare        83.0249                0.354727             0.165813
Pclass      63.8294                0.272713             0.127476
SibSp       20.9053                0.0893188            0.0417509
Parch       6.60053                0.028201             0.0131822

# QUESTION	1:	What	categories	of	passengers	were	most	likely	to	survive	the	Titanic	disaster? 

(the	mark	includes	2	points	for	the	answer	to	the	main	questions	and	2	points	for	the	detailed
explanation of the	logic	of	the	analysis)	

1) Sex=Femaie.  Almost twice as many men as women embarked but more than twice as many women survived as men so being female significantly increased how likely you were to survive.

    All passengers embarked: 
        male : 65% (577)
        female : 35% (314)
    Survivors: 
        males : 18% (109)
        females survived: 40% (233)

2) Age<30.     The avg female age was 27.9 but for survivors it was 28.7 The avg male age was 30.726645 but for survivors it was 27.8 or lower than for females.  Being younger helped men but not women.

    All passengers embarked: 
        male : 65% (577)
        female : 35% (314)
    Survivors: 
        males : 18% (109)
        females survived: 40% (233)


    All passengers embarked: 
        male : 30.7 years
        female : 27.9 years
    Survivors: 
        males : 27.8 years
        females: 28.7 years

3) Fare.  A higher fare was better but this was probably correlated with Pclass.

    The average fare was 32.204208
    The average survivors fare was 48.395408

4) Pclass=1 or 2. It looks like Pclass is a weak explanatory variable but 1st or 2nd class seems better for survival.
    
    Female survivor avg. Pclass: 1.918455
    Male survivor avg. Pclass: 2.018349

5) Fewer Siblings.
    Survivors number of siblings:
    
        No. Siblings     Survivors
        0                 61.4%
        1                 32.7%
        2                  3.8%
        3                  1.2%
        4                 0.88%

6) Fewer parents or children.

        No. Parent/Child Survivors
        0                 68.1%
        1                 19.0% 
        2                 11.7%
        3                  8.8%
        5                  0.3%

7) Embarking from Cherbourg seems to have helped--they had a higher survival rate.

        All Embarking
        
            City .        Survivors
            Southampton    72.4%
            Cherbourg      18.9%
            Queenstown      8.7%

        Survivors
        
            City .         Survivors
            Southampton    63.8%
            Cherbourg      27.4%
            Queenstown      8.8%


I also created accurate Random Forest and Gradient Boosting models in H2O and both showed that Sex, Age, Fare and PClass were most explanatory for Survival in that order.

QUESTION	2: What	other	attributes	did	you	use	for	the	analysis?	Explain	how	you	used them.	Provide	a	
complete	list	of	all	attributes	used.	à 2	points

    USed:
        PassengerId      used for counting
        Survived         used as response variable
        Pclass           used as explanatory variable
        Sex              used as explanatory variable
        Age              used as explanatory variable
        SibSp            used as explanatory variable
        Parch            used as explanatory variable
        Fare             used as explanatory variable
        Embarked         used as explanatory variable
        
    Excluded:
        Name            
        Ticket          
        Cabin   

QUESTION	3: Did	you	engineer	any	attributes?	If	yes,	explain	the	rationale	and	how	the	new	attributes	
were	used	in	the	analysis?	If	you	have	excluded	any	attributes	from	the	analysis,	provide	an	explanation	why	you	believe	they	have	to	be	excluded	(2	points)

    I did not engineer any attributes.
     
    Excluded:
        Name              unique--not explanatory          
        Ticket            mostly unqiue--not explanatory         
        Cabin             unqiue and sparsely populated--not explanatory 

QUESTION	4: How	did	you	treat	missing	values?	Provide	a	detailed	explanation	in	the	comments.	(2 points)

Some male and female passenger Age observations were NaN so I replaced them with the average age for men and women respectively.

